---
title: "Find bee SRA runs from metadata"
author: "Blake Paget"
date: '2025'
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

## Background
RNA-Seq data will be analysed to determine the expression pattern of an uncharacterised protein.
The Sequence Read Archive (SRA) maintained by NCBI will be used. The metadata will be searched for appropriate runs. In a separate publication selected runs will be downloaded and analysed.<br>

"SRA has deposited its metadata into BigQuery to provide the bioinformatics community with programmatic access to this data.
You can now search across the entire SRA by sequencing methodologies and sample attributes."<br>

A simple SQL query will be used to retrieve a table for further analysis in R. The sample attributes are an inconsistent mess and rather than trying to sort that with complex SQL queries, I'll just use R.

## Script
#### Load libraries
```{r libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(bigrquery)
library(dotenv)
library(ggplot2)
```
<br>

#### Import csv from BigQuery
```{r bq}
# Use environment variables to mask private data.
load_dot_env()

# Read the csv bq table if available, otherwise redo the SQL query.
df <- try(read.csv("bquxjob_722b10fc_1971e82ea0f.csv"))
if (inherits(df, "try-error")) {
  # BigQeury SQL query
  query <-
    "
    SELECT *
    FROM `nih-sra-datastore.sra.metadata`
    WHERE
      organism = 'Apis mellifera' AND
      consent = 'public' AND
      assay_type = 'RNA-Seq'
    "
  tb <- bq_project_query(Sys.getenv("BIGQUERY_PROJECT"), query)
  df <- bq_table_download(tb)
}
```
<br>

#### Data cleaning
```{r cleaning1}
glimpse(df)

# Any NA rows for primary key 'acc'?
is.na(df$acc) %>% table()

# Check file sizes.
is.na(df$mbytes) %>% table()
max(df$mbytes)
min(df$mbytes)

# What entries have 0 Mb?
df %>%
  filter(mbytes == 0) %>% 
  group_by(bioproject) %>% 
  count()

df %>% 
  filter(bioproject == "PRJNA477521") %>% 
  select(acc, mbases, mbytes, library_name) %>% 
  glimpse

# One bioproject with samples with blank reads? Remove from list.
df <-
  df %>% 
  filter(mbytes != 0)

nrow(df)

# Looks like the 'attributes' and 'jattr' are nested as JSON objects. 
# Not shown here, I have extracted these and compared them.
# These each contain the same sample info, e.g. 'tissue_sam_ss_dpl145': 'thoracic muscle'.
# Expand 'jattr' and add it to the df. Drop 'jattr' and 'attributes' nested columns.
```
```{r cleaning2, warning=FALSE, message=FALSE}
library(jsonlite)
```
```{r cleaning3}
parsed_jattr <- lapply(df$jattr, fromJSON)
jattr_tibble <- 
  lapply(parsed_jattr, function(x) { 
    t(cbind(names(x), as.character(unname(x))))
  }) %>% 
  lapply(function(x) {
    z <- as.data.frame(x[-1, ])
    row.names(z) <- x[1, ]
    as.data.frame(t(z))
  }) %>% 
  do.call(plyr::rbind.fill, .) %>% 
  as_tibble

jattr_df <- as_tibble(cbind(df[, !names(df) %in% c("attributes", "jattr")], jattr_tibble), .name_repair = "unique")
```
<br>

#### Data analysis

I'm interested in sectioning the RNA-seq data by caste and large body structure initially.
The keywords to look for across the expanded table are listed in the code chunk below.<br>


```{r analysis1}
# Get only the character columns.
df_chr <-
  jattr_df %>% 
  select(where(is.character))
```

```{r analysis2}
# Define keywords. Note "whole.body" is used so regex searches for any character none or more times in between
# the two words.
keywords <- c("female", "male", "drone", "queen", "worker", "nurse", "forager", "whole.body", "head", "thorax", "abdomen")
keywords <- sort(keywords)

# Grep each row of the table for the keyword, producing a list of all columns
# with the row number under the column the keyword appeared in.
# Count the number of row indices for each column (length).
keyword_list <- list()
for (keyword in keywords) {
  keyword_list[keyword] <- 
      as_tibble(sapply(df_chr, function(x) { grep(keyword, x, ignore.case = T)}) %>%
        lapply(., length) %>%
        unlist)
}

# Tidy.
keyword_find <- as.data.frame(do.call(rbind, keyword_list))
names(keyword_find) <- names(df_chr)
keyword_find <- keyword_find[ , colSums(keyword_find) != 0]
ncol(keyword_find) # the keywords appear in this many columns of the expanded metadata.
```
<br>

The table below gives the number of times any keyword appears in a column, there are 34 columns in total.

```{r analysis3}
# Summary of keyword spread.
data.frame(freq = sort(colSums(keyword_find), decreasing = T)) # most frequent columns
```
<br>

Ideally I want to find SRA runs that are clearly defined by caste:tissue or sex:tissue. This can be achieved by using grep and the & logical operator. I'll see whats available with all two-keyword combinations.<br>

```{r analysis4}
# All combinations (x2) of keywords with replacement and distinct items.
key_key <- arrangements::combinations(x = keywords, k = 2, replace = TRUE)

# Add word boundaries for grep so that male is not pulled from female or other unforeseen cases.
bounded_keys <- apply(t(key_key), 1, function(x) { paste0("\\b", x, "\\b") })

# Filter for both keywords, ignore case.
df_chr_searched <- apply(bounded_keys, 1, function(x) {
  jattr_df %>% 
    select(where(is.character)) %>% 
    filter(if_any(everything(), ~ grepl(x[1], ., ignore.case = T)) & if_any(everything(), ~ grepl(x[2], ., ignore.case = T)))
})
df_searched_count <- lapply(df_chr_searched, count)

# Combine keyword combinations and their counts.
key_key_count <- as.data.frame(cbind(key_key, unlist(df_searched_count)))

# Use df_chr_searched to get total size of all SRA runs for each combination.
acc <- lapply(df_chr_searched, select, acc)
df_gb_searched <- lapply(acc, function(x) {
  jattr_df %>% 
    semi_join(x, by = "acc") %>% 
    summarise(sum = sum(mbytes)) %>% 
    pull(sum)
})

# Add gb info to key_key_count table. 
key_key_count$gb <- round(unlist(df_gb_searched)/1000, digits = 0)

# Duplicate the counts for the reciprocal combinations to make the plots more readable.
recip_key_key_count <- cbind(V2 = key_key_count$V1, V1 = key_key_count$V2, V3 = key_key_count$V3, gb = key_key_count$gb)
key_key_count <- rbind(key_key_count, recip_key_key_count)
key_key_count <- key_key_count[!duplicated(key_key_count), ]
key_key_count$V3 <- as.numeric(key_key_count$V3)
key_key_count$gb <- as.numeric(key_key_count$gb)
```
<br>

#### Plot

```{r analysis5}
ggplot(key_key_count, aes(V1, V2)) +
  geom_tile(aes(fill = V3)) +
  geom_text(aes(label = V3)) +
  scale_fill_gradient(low = "white",
                      high = "orange",
                      name = "SRA runs") +
  labs(title = "Number of SRA runs given a combination of two keywords",
       subtitle = "Diagonal represents a single keyword") +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4, size = 12),
        axis.text.y = element_text(size = 12),
        axis.line.x = element_line(color = "black"),
        axis.line.y = element_line(color = "black"),
        panel.grid.major = element_blank(),
        plot.caption.position = "plot",
        plot.caption = element_text(hjust = 0),
        plot.title.position = "plot",
        panel.background = element_blank(),
        panel.border = element_rect(color = "grey60", fill = NA),
        legend.position = "right",
        strip.background = element_rect(color = "grey60", fill = "white"))

ggplot(key_key_count, aes(V1, V2)) +
  geom_tile(aes(fill = gb)) +
  geom_text(aes(label = gb)) +
  scale_fill_gradient(low = "white",
                      high = "lightblue",
                      name = "Size (Gb)") +
  labs(title = "Total size of SRA runs given a combination of two keywords",
       subtitle = "Diagonal represents a single keyword") +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4, size = 12),
        axis.text.y = element_text(size = 12),
        axis.line.x = element_line(color = "black"),
        axis.line.y = element_line(color = "black"),
        panel.grid.major = element_blank(),
        plot.caption.position = "plot",
        plot.caption = element_text(hjust = 0),
        plot.title.position = "plot",
        panel.background = element_blank(),
        panel.border = element_rect(color = "grey60", fill = NA),
        legend.position = "right",
        strip.background = element_rect(color = "grey60", fill = "white"))
```
<br>